src/ucp/rma/basic_rma.c

```diff
@@ -128,6 +280,98 @@ ucs_status_t ucp_get(ucp_ep_h ep, void *buffer, size_t length,
     return UCS_OK;
 }
 
+static ucs_status_t ucp_progress_get_nbi(uct_pending_req_t *self)
+{
+    ucs_status_t status;
+    size_t frag_length;
+
+    ucp_request_t *req = ucs_container_of(self, ucp_request_t, send.uct);
+
+    ucp_ep_t *ep = req->send.ep;
+    ucp_rkey_h rkey = req->send.rma.rkey;
+    uct_rkey_t uct_rkey = UCP_RKEY_LOOKUP(ep, rkey);
+
+    for (;;) {
+        frag_length = ucs_min(ep->config.max_bcopy_get, req->send.length);
+        status = uct_ep_get_bcopy(ep->uct_ep,
+                                  (uct_unpack_callback_t)memcpy,
+                                  (void*)req->send.buffer,
+                                  frag_length,
+                                  req->send.rma.remote_addr,
+                                  uct_rkey,
+                                  NULL);
+        if (ucs_likely(status == UCS_OK || status == UCS_INPROGRESS)) {
+            /* Get was initiated */
+            req->send.length -= frag_length;
+            req->send.buffer += frag_length;
+            req->send.rma.remote_addr += frag_length;
+            if (req->send.length == 0) {
+                /* Get was posted */
+                ucp_request_rma_complete(req);
+                status = UCS_OK;
+                break;
+            }
+        } else {
+            /* Error - abort */
+            break;
+        }
+    }
+
+    return status;
+}
+
+ucs_status_t ucp_get_nbi(ucp_ep_h ep, void *buffer, size_t length,
+                         uint64_t remote_addr, ucp_rkey_h rkey)
+{
+    ucs_status_t status;
+    size_t frag_length;
+
+    uct_rkey_t uct_rkey = UCP_RKEY_LOOKUP(ep, rkey);
+
+    UCP_RMA_CHECK_PARAMS(buffer, length);
+    uct_rkey = UCP_RKEY_LOOKUP(ep, rkey);
+
+    for (;;) {
```

Comment:
it is a bit similar but not code duplication. I tried to merge the two functions and it looked really ugly and over complex. This looks much cleaner.